### DISCUSSION TITLE
Questions about indentation-aware parsing for YAML

### DISCUSSION THREAD

**Discussion by martin-fleck-at** [eclipse-langium/langium#1653] [General]
Hi everyone,

First, thank you for the fantastic work you're doing! I have experience using Xtext in multiple projects and have recently started applying Langium, and I'm really impressed with its architecture and codebase. I really appreciate it.

In one of those projects, [CrossModel](https://github.com/CrossBreezeNL/crossmodel), we use Langium as the foundation for a model management server handling YAML-based models. For this, we developed a [basic indentation-based YAML lexer](https://github.com/CrossBreezeNL/crossmodel/tree/main/extensions/crossmodel-lang/src/language-server/lexer) that ignores [list item markers](https://github.com/CrossBreezeNL/crossmodel/blob/main/extensions/crossmodel-lang/src/language-server/lexer/cross-model-indentation-tokens.ts#L88) (`- `), treating them as part of the whitespace for indentation.

I was happy to see the latest master branch includes an indentation-aware lexer that's more in line with Langium’s architecture and offers customization options. I attempted to apply it to our grammar using the following custom class to handle the indentation regex while ignoring the '- ' character:

```ts
export class CrossModelTokenBuilder extends IndentationAwareTokenBuilder {
    protected readonly indentRegExp = /([ \t]+(- )?)/y;

    constructor(options?: Partial<IndentationTokenBuilderOptions>) {
        super(options);

        (this as any).indentTokenType = createToken({
            name: this.options.indentTokenName,
            pattern: this.indentMatcher,
            line_breaks: false,
        });
    }

    protected override indentMatcher: CustomPatternMatcherFunc = (text, offset, tokens, _groups) => {
        const { indentTokenName } = this.options;

        if (!this.isStartOfLine(text, offset)) {
            return null;
        }

        this.indentRegExp.lastIndex = offset;
        const match = this.indentRegExp.exec(text);
        const currIndentLevel = match?.[0].length ?? 0;
        const prevIndentLevel = this.indentationStack.at(-1)!;

        if (currIndentLevel <= prevIndentLevel) {
            return null;
        }

        this.indentationStack.push(currIndentLevel);

        const indentToken = this.createIndentationTokenInstance(
            this.indentTokenType,
            text,
            match?.[0] ?? indentTokenName,
            offset,
        );
        tokens.push(indentToken);

        return null;
    };
}
```

However, I encountered an error: "unexpected character: ->\n<- at offset: 7, skipped 1 characters." for example for this file:
![image](https://github.com/user-attachments/assets/99c2da1b-6b3c-4828-a31a-b21084b0b9c4)

This error breaks syntax highlighting. Interestingly, when I test it with a simplified version of the grammar in indentation-aware.test.ts, the error does **not** appear.

Given this, I have several questions:

1. Is this approach the right way to customize the lexer in this scenario, or is there a simpler method? If customization is needed, I’d be happy to propose a PR to make this easier.

2. Why might syntax highlighting be broken? Could it be related to the parsing error?

3. Any insights on why the "unexpected character" error occurs or suggestions for debugging this effectively? Are there any visualization tools available that could help here?

4. Is there a nightly build of Langium? I didn't find one on npm under the next tag, so I manually copied the indentation-aware.ts file. Could missing other updates cause these issues?

5. It seems that `whitespaceTokenName` overrides the token with a token defined by `whitespaceRegExp`. Does this conflict with existing WS rules?

6. Does Langium discourage PRs for typos? I noticed a typo in `ignoreIndentationDelimeters` (should be `ignoreIndentationDelimiters`). I'm aware of backward compatibility concerns once the new release is out with the type, so I'd like to contribute a fix but wasn't sure as some projects do not like that.

Any help or insight (even on a single question) would be greatly appreciated! If YAML is an interesting use case for Langium, we're also open to contributing this work via a PR.

Thanks in advance!

**Comment by msujew** [comment]
cc @aabounegm Since you've worked quite a bit in this area, can you help? :)

**Reply by aabounegm** [reply]
Yes sure, sorry for the delay, writing a reply now

**Comment by aabounegm** [comment]
Hi @martin-fleck-at, let me try to answer the questions one by one:
1. Generally, this is the right way to override/extend the functionality of the lexer, but in this specific case it would be much simpler to just override the `whitespaceRegExp` property with your RegEx that matches whitespace and `-`s. Just make sure not to forget the `y` flag. If the RegEx is not flexible enough, you can override the `matchWhitespace` method instead.
2. The syntax highlighting is most likely broken due to the use of `this_string_does_not_matter_spaces#$%^&*((` as the terminal pattern. The TextMate generator uses the source Langium file and not the overridden token definitions as the source for the generated grammar. As can be seen in [textmate-generator.ts:L77](https://github.com/eclipse-langium/langium/blob/main/packages/langium-cli/src/generator/highlighting/textmate-generator.ts#L77), it uses `GrammarUtils.isCommentTerminal`, which defines a comment terminal as any hidden terminal that is not whitespace. The string you used satisfies this condition, and though I'm not sure what `RegExpUtils.getTerminalParts` would return here (need to test it myself), I'm suspecting this is the source of the problem. You should use token definitions that resemble actual `SPACES` and `NEWLINE`s, and note in a comment that it actually doesn't matter (for the actual parser).
This should not be related to the parse error, which is the next point:
3. I noticed in your [terminals.langium:L31](https://github.com/CrossBreezeNL/crossmodel/blob/main/extensions/crossmodel-lang/src/language-server/terminals.langium#L31) that you wrote the `NEWLINE` terminal definition doesn't matter, but actually this is not a token that gets overridden by the `IndentationAwareTokenBuilder`. The definition actually matters for this token. I use `/[\r\n]+/` for it. I don't know about visualization tools, but when debugging the token builder I personally add something like `console.log('tokens', result.tokens.map(tk => [tk.tokenType.name, tk.image]))` to the `tokenize` function of the lexer just to check what tokens are lexed.
4. I'm generally not part of the release process, so @msujew could probably help more here, but I know there is a `3.2.0-next` release on npm. Also, you can install the package [using GitHub URL](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#github-urls) to get the version on GitHub rather than npm.
5. The reason we override this token is to ensure that it does not include new lines (because most users are probably used to just writing `\s+` for the `WS` token, which would break the indentation matching. I'm not sure if the choice of reusing `this. whitespaceRegExp` is better than hard-coding `[ \t]+`, perhaps your use case would better clarify this.
6. Again, I'm just a contributor myself, but I assume such PRs are welcome as I have personally submitted one to the `langium-website` repo previously :). Since v3.2.0 is not released yet as stable, I believe it is a good idea to fix the typo before it gets released.

Hope this helps, let me know if you have any follow-up questions!

**Reply by msujew** [reply]
Regarding 4 - we don't have a nightly. I usually create a few release candidates to check for regressions before a new release.

Regarding 6 - All PRs are appreciated :)

**Reply by martin-fleck-at** [reply]
@aabounegm Thank you very much for the detailed answer, I really appreciate it! I got a bit further with it and the error I was seeing was indeed related to the newline-handling - I somehow messed it up by not including the newlines for all platforms (I use `/[/\r\n|\r|\n/]+/` now)! So that is already a huge help! I'm still facing some issues with whitespaces destroying my parser and I'll need some time to dig a bit deeper on my end but I'll definitely report back either way within the next few days.

@msujew Perfect, then I'll open a PR to fix the typos within the next 1-2 days, just want to make sure there is nothing else that could easily be changed within the same PR. Thank you also for your quick response and pushing a "nightly" release ;-)

**Reply by aabounegm** [reply]
Glad to be of help. Waiting for your updates regarding the parser :)

> just want to make sure there is nothing else that could easily be changed within the same PR

IMHO, multiple tiny PRs are better than one PR with unrelated changes because they are easier to review and merge, but I'm not sure whether that's what the maintainers here prefer 😅

**Reply by martin-fleck-at** [reply]
So I've made enough progress to report back: I believe most issues were actually my fault by not properly considering whitespaces or newlines and it is actually enough to adapt the `matchWhitespace` method as you suggested. 

I did notice a few minor issues with the token builder:
- My completion suggestions are no longer picked up properly because the token builder fills up the partially parsed text with the remaining dedents, making the way I query what the next feature should be invalid. I introduced a boolean flag into the token builder to turn off that behavior for the completion parser which re-uses the same token builder as the "regular" parser. This may not be the nicest solution but it works.
- When creating the token types, we cannot easily override the pattern as it is directly called. Calling it with `pattern: (text, offset, tokens, groups) => this.indentMatcher(text, offset, tokens, groups)` would fix that problem.
- Reporting errors is something that is already marked a comment in the token builder. I wonder if we could not simply use a custom token type for that. I didn't find any good way in chevron to communicate that and throwing an error has the downside of not only crashing the server (which could be prevented with a catch) but also that the whole process aborts.

I have two files in my draft PR to reflect that:
- [cross-model-indentation-aware.ts](https://github.com/CrossBreezeNL/crossmodel/pull/64/commits/8a21a269f450cef5e1b586fe08117a1a69235f34#diff-2b8a2ff48730f690063e9789abd4043823dabecc8e92edd833c6610413df2e5e) containing some of the fixes.
- [indentation-aware.ts](https://github.com/CrossBreezeNL/crossmodel/pull/64/commits/8a21a269f450cef5e1b586fe08117a1a69235f34#diff-94fc26517778b187ffd1e7f426b88b29ed7606630d4418951dc0a37a323ed1e7) which contains some of the fixes and extensions I did in the base class.

I'll open some separate PRs tomorrow where we could discuss whether the problems I faced were generic enough and the solutions are actually suitable. In any case, thank you very much again for the great code and the quick response, highly appreciated.

**Reply by aabounegm** [reply]
About the completion suggestion, I haven't yet tried to implement my own, so I can't help much till I check how it works, but it seems a bit strange to me that the remaining dedents would do this.

> we cannot easily override the pattern as it is directly called

Right, looks like I forgot this nuance between class methods and normal fields with functions. It should be changed to a normal method to allow overriding and calling the `super` method. I submitted #1664 now to fix this

> Reporting errors is something that is already marked a comment in the token builder. I wonder if we could not simply use a custom token type for that.

A custom token type seems like a hack, but it should work. I think it's better to start a separate discussion for a proper solution for reporting diagnostics from the parser (because there can be use cases when a token is valid but should still issue a warning, such as when mixing spaces and tabs in indentation)

**Reply by martin-fleck-at** [reply]
I see that you already opened PRs for the method handling and the typos! Great! As to how we can report the errors on tokens, it seems to me that the chevrotain way is to use [custom payloads](https://chevrotain.io/docs/guide/custom_token_patterns.html#custom-payloads) in the custom matchers that end up as a [property on the token](https://github.com/Chevrotain/chevrotain/blob/4272462f2c88846357dc95dfdd5684d87c153c4f/packages/chevrotain/src/scan/lexer_public.ts#L919) so we could evaluate them in our lexer wrapper after calling the lexer of Chevrotain and before returning the lexer result.

However, I noticed that in our indentation-based implementation our custom pattern matchers always return `null`, what is the reason for that?

**Reply by msujew** [reply]
See https://github.com/eclipse-langium/langium/blob/f60cfeb6842ee972d81623b21d67bd54ed5029e0/packages/langium/src/parser/indentation-aware.ts#L303-L308

It might make sense to simply return the token instead of pushing it to the token list.

**Reply by aabounegm** [reply]
@martin-fleck-at I also thought about using custom payloads, but it seems a bit inefficient to go through all tokens in the lexer to check if they have a payload (also, they're not really type-safe). Additionally, that doesn't handle the case of emitting warnings, which the lexer itself doesn't yet support.

@msujew I agree, I think it was a case of copy-pasting from the [Chevrotain example](https://github.com/chevrotain/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js) and I forgot to modify it. It should work for indentation, but for dedent we might need to add more than one, so adding some manually and only the last one implicitly via `return` might be a bit confusing

**Reply by martin-fleck-at** [reply]
@aabounegm Yeah, efficiency is definitely a point that we need to consider but since chevrotain is a black box for us, I'm not sure there is a way to avoid going through the return result of the lexer. As for type safety, we surely could defined an interface and some type checks for us so I'm less worried about that. Regarding warnings, we could introduce a custom "sub-type" of ILexerError (e.g., ILexerWarning) when we translate our token payload. Those errors (and warnings) are later tranlsated [into diagnostics](https://github.com/eclipse-langium/langium/blob/f60cfeb6842ee972d81623b21d67bd54ed5029e0/packages/langium/src/validation/document-validator.ts#L99) where we can properly distinguish between the two types, right?

**Reply by aabounegm** [reply]
Even if we consider Chevrotain to be a black box, the Token Builder and Lexer in Langium are not, and using a custom matcher function, we can always add any logic we want when a token is matched, similar to how the indentation token matcher updates the `indentationStack`.

For instance, I can imagine the following example hypothetical scenario (which would be nice if supported in the future):
1. Langium adds support for a `deprecated` modifier for tokens in the grammar
2. When a token has that modifier, the Chevrotain token generated for it would do something like:
    ```ts
    const originalRegex = token.pattern;
    token.pattern = (text, offset) => {
      originalRegex.lastIndex = offset;
      const match = originalRegex.exec(text);
      if (match) { this.warnings.push(/* Some info */); } // or maybe a more generic "this.diagnostics"
      return match;
    }
    ```
3. The lexer would collect these warnings/errors from the token builder
4. The parser would also pass them along in `ParseResult`
5. They would all be handled in the function you linked

At least this is how I imagine it could be implemented (the first 2 points are specific to this example, but the rest is generic and can possibly be implemented in Langium)

**Reply by aabounegm** [reply]
As for token payloads, I was actually curious if they can be accessed somehow in the AST/CST, but that's a different discussion (I've been meaning to start a discussion, but haven't gotten around to it just yet)

**Reply by msujew** [reply]
We store the `IToken` instance on the CST, so it's definitely possible to retrieve the payload from there.

**Reply by martin-fleck-at** [reply]
I totally agree with what you said and I think we only diverge in the implementation of step 2, where I'd much rather use the `payload` to communicate warnings and errors as opposed to tracking them as a list but in the end that probably does not matter if we can guarantee that we always finish all the tokenizing before and generate all the errors and warnings before another call is made, i.e., we need to ensure there are no syncing issues where the error and warning collections might be accessed wrongly. If we can make sure that this is the case, those lists are probably more efficient than the payload mechanism which still has the problem that in the dedent-case we cannot would end up with a weird behavior where we return one match for one token and still need to manually add the others if we have more than one. In any case, I'm very happy with either solution.

**Reply by aabounegm** [reply]
@msujew I cannot seem to find it, are you talking about the [`CstNode` interface](https://github.com/eclipse-langium/langium/blob/f60cfeb6842ee972d81623b21d67bd54ed5029e0/packages/langium/src/syntax-tree.ts#L227)?

**Reply by aabounegm** [reply]
@martin-fleck-at I believe parsing is fulling synchronous, so there shouldn't be any race conditions.
The benefit of a separate list would be that it's more efficient (no need to go through all tokens again and check if they have a payload), we wouldn't lose type-safety for these diagnostics (`IToken.payload` has type `any), and also that users would still be able to use the payload for their own actual custom payloads. And thank you for reminding me of the added benefit that you don't actually need to match the token to be able to add a message

**Reply by martin-fleck-at** [reply]
Maybe I should also start a separate discussion for this as well but since it seems a release is upon us soon, I want to bring it up one more time: For me, the completion provider is no longer working if we finalize the parsed text with dedents as this concludes the YAML document. In particular, [finding the next features](https://github.com/eclipse-langium/langium/blob/f60cfeb6842ee972d81623b21d67bd54ed5029e0/packages/langium/src/lsp/completion/completion-provider.ts#L201) always yields an empty array as there is nothing after the document is closed with the final dedent. I'd assume it might be similar for other languages where dedents actually close out a block so I'm not sure we want to handle this on a framework level where for completion parsing we deactivate the dedent auto-fill. Any thoughts on that? I'd be happy to provide a branch/PR for this and also the other topic, if that helps.

**Reply by martin-fleck-at** [reply]
> @martin-fleck-at I believe parsing is fulling synchronous, so there shouldn't be any race conditions. The benefit of a separate list would be that it's more efficient (no need to go through all tokens again and check if they have a payload), we wouldn't lose type-safety for these diagnostics (`IToken.payload` has type `any), and also that users would still be able to use the payload for their own actual custom payloads. And thank you for reminding me of the added benefit that you don't actually need to match the token to be able to add a message

Truly a lot of benefits. If we are fully synchronous that is definitely the way to go! Do you want to open a PR for this or should I do it? I'd probably have some time tomorrow or Monday next week.

**Reply by aabounegm** [reply]
> other languages where dedents actually close out a block

This reminded me to ask, in normal languages that use `{}` (for example) to denote blocks, does the completion provider work if the cursor is placed before the closing `}`? Because dedent tokens and `}` should behave very similarly. If it works with `}`, then there might be something wrong with the dedent token itself. I noticed now that when creating the token, I used `match?.[0] ?? dedentTokenName` for the image, which might be messing with position calculations that depend on `match.length`. When writing that initially, I only considered that `dedentTokenName` might be more useful for debugging than an empty string. Could you perhaps try changing that to `''` and see if it changes anything?

> Do you want to open a PR for this or should I do it?

I don't mind either way. You can open a PR if you like, I'm a bit busy tomorrow, but I can't promise I won't get bored during the weekend and implement it before you do 😀

**Reply by msujew** [reply]
@aabounegm I was talking about `LeafCstNode` but noticed that it only exposes the tokenType

https://github.com/eclipse-langium/langium/blob/f60cfeb6842ee972d81623b21d67bd54ed5029e0/packages/langium/src/syntax-tree.ts#L265

**Reply by aabounegm** [reply]
Yeah, I faced this issue when implementing my hover provider and wanted to restrict the hover range to just a part of the node, but that part is a plain string and not a nested node, so I didn't know how to easy get the range of that string

**Reply by msujew** [reply]
> This reminded me to ask, in normal languages that use {} (for example) to denote blocks, does the completion provider work if the cursor is placed before the closing }?

Yes, the issue is that the indentation aware lexer adds synthethic tokens that aren't really there in the input. A normal lexer simply wouldn't find the `}` and therefore emit no token of that type.

**Reply by aabounegm** [reply]
What I meant was what is supposed to happen if completion is manually triggered when there is already a `}` after it?
Keep in mind that IDEs usually auto-close the bracket pair, so the `}` would probably be there anyway.

**Reply by msujew** [reply]
The completion only takes token up to the cursor position into account - all other tokens are ignored. Therefore, the `}` token simply isn't visible to the completion. This is the issue with the synthetic dedent tokens - they're technically "after" the cursor position and therefore mess with the completion calculation.

**Reply by aabounegm** [reply]
Got it. Could the problem be mitigated by adding a `filter` to this line:
https://github.com/eclipse-langium/langium/blob/f60cfeb6842ee972d81623b21d67bd54ed5029e0/packages/langium/src/lsp/completion/completion-provider.ts#L224
like this:
```ts
const leftoverTokens = [...tokens].splice(parserResult.tokenIndex).filter(token => token.startOffset < offset);
```
?
(or perhaps a more efficient variant of `filter` that starts from the end and stops with the first matching element, since order is guaranteed to be monotonically increasing)

**Reply by martin-fleck-at** [reply]
I'll try this out. I'm working currently on a PR for the diagnostics and will then check whether we can fix the auto-completion provider in a better way. My alternative approach to this would have been to allow an additional option to the `tokenize` method of the lexer that tells us whether we parse a full or partial document. But if we can make the filter more clever, maybe this is a better idea.

**Reply by martin-fleck-at** [reply]
I just tested this and it does not work. In my case, let's assume this is the full document:

```
entity:
    id: CalcAge
    name: "CalcAge"
    attributes:
      - id: BirthDate
        name: "BirthDate"
        datatype: "Integer"
```

And I put the cursor before the `dataType`, the following text is considered for completion parsing:

```
entity:
    id: CalcAge
    name: "CalcAge"
    attributes:
      - id: BirthDate
        name: "BirthDate"
        <cursor positition at offset 122>
```

As offset I get 122 (the cursor position) but the tokens of dedent are still in the way to properly detect the next feature, unfortunately.

**Reply by martin-fleck-at** [reply]
I opened https://github.com/eclipse-langium/langium/pull/1669 as an option to properly distinguish between the two modes.

**Comment by aabounegm** [comment]
@msujew About the `LeafCstNode` exposing the token type, it seems straightforward to me to change it to store the whole token (unless I'm missing something obvious), I can prepare a PR for that. However, I still don't quite understand how to easily access the `LeafCstNode` of a token.
For example, let's say we're using the Hello World language that generates this AST:
```ts
interface Person extends AstNode {
    $type: 'Person';
    name: string;
}
```
then, let's say in my hover provider, I have:
```ts
if (isPerson(node)) {
    const range = ?? // How can I extract the range of just the `name` property, not the whole node? 
    return { contents: `A human named ${node.name}`, range };
}
```
I would like the hover range to be only on the name, how can I get this information easily? It would be great if there was an associated `nameToken: IToken` in the `Person` interface.

**Reply by martin-fleck-at** [reply]
I really can't help with this, but the only thing I noticed while looking at the diagnostics report: We need to ensure that the `IToken` can be properly serialized/deserialized as part of the hydration process that is used for async parsing.

**Reply by aabounegm** [reply]
That's a good point. Looking at the [Chevrotain IToken docs](https://chevrotain.io/documentation/6_3_0/interfaces/itoken.html), it seems that all properties have primitive types except for `tokenType` (which is already handled in the hydrator), and the `payload` which can have any type, but I think it's safe to assume that it would very rarely (if ever) be used with a non-serializable type.

**Reply by msujew** [reply]
I don't think that storing the whole token instance is necessary. We can already perfectly find the range for any property/keyword using:

https://github.com/eclipse-langium/langium/blob/51d99a66d8cddff2729a35d6cee62eea1203923d/packages/langium/src/utils/grammar-utils.ts#L111-L134

That's how we do all the LSP stuff, like identifying where the `name` of an element is stored in a document or where to apply diagnostics given an `AstNode` and the name of a property.

**Reply by aabounegm** [reply]
Oh, I see, I should have read the source code more closely 😅. I think it would help if the documentation had an "API reference" section, perhaps generated from the JSDoc
